### PR TITLE
Adding aci-operator-config config-map watcher to ACI containers operator 

### PR DESCRIPTION
**Pull request by abhis2112**
1. The Watcher watches the aci-operator-config config-map
2. Change in config-map will handle the respective change in acicontainersoperator CR.
3. This in turn apply the latest config to the cluster.
4. This case will handle the update done by acc-provision-operator during the update of aci-operator-configmap
   and passing the change to acicontaineroperator CR.

### CODE REVIEW

File: pkg/controller/acicontainersoperator.go [hunk 1]
```diff
@@ -329,6 +350,35 @@ func NewAciContainersOperator(
 			}
 		},
 	})
+
+	config_map_informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			log.Debug("The ConfigMap key: ", key)
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:357]: Assuming this event does not happen often, we should probably log this at Info level. Also, can you make the log a little more descriptive. One should not have to go back and look in the code to understand the context of the log statement.

File: pkg/controller/acicontainersoperator.go [hunk 2]
```diff
@@ -329,6 +350,35 @@ func NewAciContainersOperator(
...
+			current_Obj := obj.(*corev1.ConfigMap)
+			if current_Obj.Name == "aci-operator-config" {
+				if err == nil {
+					config_map_queue.Add(key)
+				}
+			}
+
+		},
+		UpdateFunc: func(prevObj, currentObj interface{}) {
+			log.Debug("In UpdateFunc for ConfigMap")
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:367]: Info level?

File: pkg/controller/acicontainersoperator.go [hunk 3]
```diff
@@ -329,6 +350,35 @@ func NewAciContainersOperator(
...
+			current_Obj := currentObj.(*corev1.ConfigMap)
+			if current_Obj.Name == "aci-operator-config" {
+				controller.handleConfigMapUpdate(prevObj, currentObj, config_map_queue)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			log.Debug("Deleted ConfigMap key is :", key)
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:375]: Info level?

File: pkg/controller/acicontainersoperator.go [hunk 4]
```diff
@@ -383,6 +433,73 @@ func (c *Controller) handledaemonsetUpdate(oldobj interface{}, newobj interface{
...
+	old_cm := oldobj.(*corev1.ConfigMap)
+	new_cm := newobj.(*corev1.ConfigMap)
+
+	if !reflect.DeepEqual(old_cm.Data, new_cm.Data) {
+		key, err := cache.MetaNamespaceKeyFunc(newobj)
+		if err == nil {
+			queue.Add(key)
+		}
+	} else {
+		log.Info("Config-Map is intact for ", new_cm.Name)
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:494]: intact -> unchanged

File: pkg/controller/acicontainersoperator.go [hunk 5]
```diff
@@ -383,6 +433,73 @@ func (c *Controller) handledaemonsetUpdate(oldobj interface{}, newobj interface{
...
+	}
+}
+
+func (c *Controller) handleConfigMapDelete(obj interface{}) bool {
+	log.Infof("aci-containers-operator config-map Deleted")
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:499]: nit: but why is "D" upper case here??

File: pkg/controller/acicontainersoperator.go [hunk 6]
```diff
@@ -413,7 +546,7 @@ func (c *Controller) CreateAciContainersOperatorObj() *operators.AciContainersOp
 
 func (c *Controller) CreateAciContainersOperatorCR() error {
 	log.Info("Reading the aci-operator-config Config Map providing CR")
-	rawSpec, err := c.ReadConfigMap("aci-containers/aci-operator.conf")
+	rawSpec, err := c.ReadConfigMap("/usr/local/etc/aci-containers/aci-operator.conf")
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:549]: Here and elsewhere below - why is this change needed? Making the file path absolute would mean that we have to be really sure that it remains the same in all situations. Would that be the case? And even if that were to be the case, this prefix should probably be pre-defined somewhere else and used here.

File: pkg/controller/acicontainersoperator.go [hunk 7]
```diff
@@ -401,6 +518,22 @@ func (c *Controller) ReadConfigMap(field string) ([]byte, error) {
...
+	if err != nil{
+		log.Error(err)
+		return err
+	}
+	err = ioutil.WriteFile( field, data_byte, 0777)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	log.Debug("Config-Map is:  ", (string(data_byte)))
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:533]: I think we should put an info level log statement stating that we are writing the config-map (name?). The actual contents of the map can be at debug level, but if the config is base-encoded perhaps there is no need to log it.

**Review by snaiksat — CHANGES_REQUESTED** [review]

**Comment by snaiksat** [comment]
Also, need to run gofmt...

File: pkg/controller/acicontainersoperator.go (hunk 7 above)
Comment (abhis2112) [reply pkg/controller/acicontainersoperator.go:535]: Yes the config-map is big base encoded log.  I will change the debug to info level with config map name

File: pkg/controller/acicontainersoperator.go [hunk 8]
```diff
@@ -329,6 +353,35 @@ func NewAciContainersOperator(
 			}
 		},
 	})
+
+	config_map_informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			log.Info("In AddFunc for Config-Map with key: ", key)
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:360]: can we put the name of the configmap in the log message? If someone reads this log message in the log, he has to go back to code and check which configmap we are talking about...

File: pkg/controller/acicontainersoperator.go [hunk 9]
```diff
@@ -329,6 +353,35 @@ func NewAciContainersOperator(
...
+			current_Obj := obj.(*corev1.ConfigMap)
+			if current_Obj.Name == "aci-operator-config" {
+				if err == nil {
+					config_map_queue.Add(key)
+				}
+			}
+
+		},
+		UpdateFunc: func(prevObj, currentObj interface{}) {
+			log.Info("In UpdateFunc for ConfigMap")
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:370]: ditto

File: pkg/controller/acicontainersoperator.go [hunk 10]
```diff
@@ -329,6 +353,35 @@ func NewAciContainersOperator(
...
+			current_Obj := currentObj.(*corev1.ConfigMap)
+			if current_Obj.Name == "aci-operator-config" {
+				controller.handleConfigMapUpdate(prevObj, currentObj, config_map_queue)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			log.Info("Deleted ConfigMap key is :", key)
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:378]: ditto

File: pkg/controller/acicontainersoperator.go [hunk 11]
```diff
@@ -379,10 +432,77 @@ func (c *Controller) handledaemonsetUpdate(oldobj interface{}, newobj interface{
...
+	}
+
+	log.Info("Reading current aci-operator-config configmap")
+	rawSpec, err := c.ReadConfigMap("/tmp/aci-operator.conf")
+	if err != nil {
+		return true
+	}
+
+	obj := c.CreateAciContainersOperatorObj()
+	log.Info("Unmarshalling the ConfigMaps...")
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:463]: name of configmap...also notice how sometimes you use "Config-map" and then sometimes you say "ConfigMaps"...admittedly a nit, but there has to be consistency in conventions at least in the same module...

File: pkg/controller/acicontainersoperator.go [hunk 12]
```diff
@@ -379,10 +432,77 @@ func (c *Controller) handledaemonsetUpdate(oldobj interface{}, newobj interface{
...
+		_, err = c.Operator_Clientset.AciV1alpha1().AciContainersOperators(os.Getenv("SYSTEM_NAMESPACE")).
+			Update(context.TODO(), acicnioperator, metav1.UpdateOptions{})
+		if err != nil {
+		}
+	}
+	return false
+}
+
+func (c *Controller) handleConfigMapUpdate(oldobj interface{}, newobj interface{}, queue workqueue.RateLimitingInterface) {
+	log.Info("In Config-Map update handler ")
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:487]: name of configmap

File: pkg/controller/acicontainersoperator.go [hunk 13]
```diff
@@ -401,6 +521,22 @@ func (c *Controller) ReadConfigMap(field string) ([]byte, error) {
...
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	err = ioutil.WriteFile(field, data_byte, 0777)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	log.Info("Writing the Config-Map: ",data.Name," in the file")
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:536]: There is no write happening after this log statement. Either the log statement should be "Wrote..." or we should move this log statement to before the write is done...

File: pkg/controller/acicontainersoperator.go [hunk 14]
```diff
@@ -109,6 +112,9 @@ const aciContainersController = "aci-containers-controller"
 const aciContainersHostDaemonset = "aci-containers-host"
 const aciContainersOvsDaemonset = "aci-containers-openvswitch"
 
+var Aci_operator_config_path = "/usr/local/etc/aci-containers/aci-operator.conf"
```
Comment (snaiksat) [inline pkg/controller/acicontainersoperator.go:115]: while this addresses the earlier comment about predefining the strings, I am still not clear why this path has to be an absolute path whereas earlier it was not...

**Review by snaiksat — CHANGES_REQUESTED** [review]

**Comment by snaiksat** [comment]
And gofmt is still failing...

File: pkg/controller/acicontainersoperator.go (hunk 8 above)
Comment (abhis2112) [reply pkg/controller/acicontainersoperator.go:360]: This can add an extra line of code to fetch the name of the config-map. It can be done if really needed. However in subsequent call in handler we mention which config-map

File: pkg/controller/acicontainersoperator.go (hunk 14 above)
Comment (abhis2112) [reply pkg/controller/acicontainersoperator.go:115]: Earlier also it was. The function ReadConfigMap is called like below 

**rawSpec, err := c.ReadConfigMap("aci-containers/aci-operator.conf")**

and inside the function, it is coded with **/usr/local/etc**

**func (c *Controller) ReadConfigMap(field string) ([]byte, error) {
	raw, err := ioutil.ReadFile("/usr/local/etc/" + field)**

So eventually it becomes absolute and we are not changing this location as we mention this location in volume mount in deployment file

**Review by snaiksat — APPROVED** [review]
